File: 309-best-time-to-buy-and-sell-stock-with-cooldown/best-time-to-buy-and-sell-stock-with-cooldown.java
class Solution {
    public int maxProfit(int[] prices) {
        int[][] dp = new int[prices.length + 2][2];

        for(int i = prices.length - 1 ; i >= 0 ; i--){
            for(int buy = 1 ; buy >= 0 ; buy--){
                if(buy == 1){
                    dp[i][buy] = Math.max(dp[i+1][1] , dp[i+1][0] - prices[i]);
                } else{
                    dp[i][buy] = Math.max(dp[i+1][0] , dp[i+2][1] + prices[i]);
                }
            }
        }

        return dp[0][1];
    }
}

File: 309-best-time-to-buy-and-sell-stock-with-cooldown/best-time-to-buy-and-sell-stock-with-cooldown.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        dp = [[0,0] for _ in range(n+2)]

        for i in range(n-1 , -1 , -1):
            for canBuy in range(2):
                if canBuy == 1:
                    dp[i][canBuy] = max(
                        dp[i+1][0] - prices[i],
                        dp[i+1][1]
                    )
                else:
                    dp[i][canBuy] = max(
                        dp[i+2][1] + prices[i],
                        dp[i+1][0] 
                    )
        
        return dp[0][1]